src/System.Runtime.Analyzers/Core/InstantiateArgumentExceptionsCorrectly.cs

```diff
@@ -1,59 +1,189 @@
+                        });
+                });
+        }
+
+        private void AnalyzeObjectCreation(
+            OperationAnalysisContext context, 
+            ISymbol owningSymbol,
+            ITypeSymbol argumentExceptionType, 
+            ITypeSymbol stringType)
+        {
+            var creation = (IObjectCreationExpression)context.Operation;
+            if (!creation.ResultType.Inherits(argumentExceptionType))
+                return;
+          
+            if (creation.ConstructorArguments.Length == 0)
+            {
+                if (HasMessageOrParameterNameConstructor(creation.ResultType, stringType))
+                {
+                    // Call the {0} constructor that contains a message and/ or paramName parameter
+                    creation.Syntax.CreateDiagnostic(NoArgumentsRule, creation.ResultType);
+                }
+            }
+            else
+            {
+                foreach (var argument in creation.ConstructorArguments)
+                {
+                    if (argument.Parameter.Type != stringType)
+                        continue;
+                 
+                    string value = argument.Value.ConstantValue as string;
+                    if (value == null)
+                        continue;
+
+                    CheckArgument(owningSymbol, creation.ResultType, argument.Parameter, value, context);
+                }
+            }
+        }
+
+        private void CheckArgument(
+            ISymbol targetSymbol,
+            ITypeSymbol constructorType, 
+            IParameterSymbol parameter, 
+            string stringArgument, 
+            OperationAnalysisContext context)
+        {
+            bool matchesParameter = MatchesParameter(targetSymbol, stringArgument);
+            DiagnosticDescriptor descriptor = null;
+
+            if (IsMessage(parameter) && matchesParameter)
+            {
+                descriptor = IncorrectMessageRule;
+            }
+            else if (IsParameterName(parameter) && !matchesParameter)
+            {
+                descriptor = IncorrectParameterNameRule;
+            }
+
+            if (descriptor != null)
+            {
+                // Method {0} passes parameter name '{1}' as the '{2}' argument to a {3} constructor. [...]
+                context.ReportDiagnostic(
+                    context.Operation.Syntax.CreateDiagnostic(
+                        descriptor,
+                        targetSymbol,
+                        parameter.Name,
+                        constructorType));
+            }
+        }
+
+        private static bool IsMessage(IParameterSymbol parameter)
+        {
+            return parameter.Name == "message";
+        }
+
+        private static bool IsParameterName(IParameterSymbol parameter)
+        {
+            return parameter.Name == "paramName"
+                || parameter.Name == "parameterName"
+                || parameter.Name == "argumentName";
```

Comment:
`ArgumentException` does not seem to have any constructor with a parameter named `argumentName`. I see that FXCop [implementation](http://index/UsageRules/R/74502e3211b70ebb.html) has similar code, but we should probably fix it here?